src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+  .split('(').join('%28')
+  .split(')')
+  .join('%29')
+  .split('*')
+  .join('%2A')
+  .split('+')
+  .join('%2B')
+  .split(',')
+  .join('%2C')
+  .split(':')
+  .join('%3A');
+
+/**
+ * @param {string} input - id that requires checkdigit
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @return {number} the checkDigit of the parameter
+ */
+const getCheckDigit = (input, throwError) => {
+  try {
+    const array = input.split('').reverse();
+    let total = 0;
+    let i = 1;
+    array.forEach((number) => {
+      const n = parseInt(number, 10);
+      if (i % 2 === 0) {
+        total += n;
+      } else {
+        total += n * 3;
+      }
+      i += 1;
+    });
+    return Math.ceil(total / 10) * 10 - total;
+  } catch (e) {
+    if (throwError) {
+      throw new Error(`There was a problem during the check digit operation of ${input}`);
+    }
+    return 0;
+  }
+};
+
+/**
+ * it adds the check sum at the end of the input and return a zero-padded result with a pre-defined
+ * length
+ * @param {string} raw - the string to convert
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @param length - the length of the output string
+ * @return {string} the updated string
+ */
+export const addCheckDigitAndZeroPad = (raw, throwError, length = 14) => {
+  let newRaw = raw;
+  newRaw += getCheckDigit(newRaw, throwError);
+  while (newRaw.length < length) {
+    newRaw = `0${newRaw}`;
+  }
+  return newRaw;
+};
+
+/**
+ * Converts any standard URI conveying a GS1 Key in Canonical GS1 DL URI.
+ * @param originalUri - any URI to be used in EPCIS events that convey a GS1 key, i.e. EPC URIs, EPC
+ * Class URIs, EPC ID Pattern URIs, or GS1 Digital Link URIs.
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @return a corresponding, constrained version of a canonical GS1 Digital Link URI, i.e. with the
+ * lowest level of identification and without CPV/query string.
+ */
+export const dlNormalizer = (originalUri, throwError = true) => {
+  let uri = originalUri;
+
+  if (!uri.includes('.')) {
+    return uri;
+  }
+
+  const partition = uri.indexOf('.');
+
+  // EPC URIs
+  if (uri.match('^urn:epc:id:sgtin:(([0-9]{6}.[0-9]{7})|([0-9]{7}.[0-9]{6})|([0-9]{8}.[0-9]{5})|([0-9]{9}.[0-9]{4})|([0-9]{10}.[0-9]{3})|([0-9]{11}.[0-9]{2})|([0-9]{12}.[0-9]{1})).(%2[125-9A-Fa-f]|%3[0-9A-Fa-f]|%4[1-9A-Fa-f]|%5[0-9AaFf]|%6[1-9A-Fa-f]|%7[0-9Aa]|[!\')(*+,.0-9:;=A-Za-z_-]){1,20}$')) {
+    const gs1CompanyPrefix = uri.substring(17, partition);
```

Comment:
regex capture groups could help avoid using lots of 'magic numbers' here.

Reply:
Indeed, but to be honest, I just followed the existing implementation of the [algorithm](https://github.com/RalphTro/epcis-event-hash-generator/blob/master/epcis_event_hash_generator/dl_normaliser.py). Therefore, I did exactly like them for the implementation of this function.

Regex could help avoid 'magic numbers' but I hadn't to calculate them so I implemented it quite fast.